### PR TITLE
Feature/team crud

### PR DESCRIPTION
**Pull request by zhamza372**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -29,4 +30,11 @@ def scope_current_company
   ensure
     Current.reset
   end
+
+  rescue_from CanCan::AccessDenied do |exception|
```
Comment (bk-az) [inline app/controllers/application_controller.rb]: remove

File: app/controllers/teams_controller.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+# Team Controller
+class TeamsController < ApplicationController
+  load_and_authorize_resource find_by: :sequence_num
+
+  def index
+    @teams = @teams.page(params[:page]).per_page(PER_PAGE)
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: `@teams = @teams.includes(:created_by, :lead).page(params[:page]).per_page(PER_PAGE)`

File: app/controllers/teams_controller.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    @users = User.all
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: remove

File: app/controllers/teams_controller.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def show
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def new
+    @users = User.all
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: token input

File: app/controllers/teams_controller.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def edit
+    @users = User.all
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: token input

File: app/controllers/teams_controller.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+  def create
+    @team.created_by_id = current_user.id
+    if @team.save
+      flash[:notice] = t 'team.created'
+      redirect_to @team
+    else
+      errors = @team.errors.full_messages.join(', ')
+      flash[:error] = errors
+      @users = User.all
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: remove

File: app/controllers/teams_controller.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def update
+    if @team.update(team_params)
+      flash[:notice] = t 'team.updated'
+      redirect_to @team
+    else
+      errors = @team.errors.full_messages.join(', ')
+      flash[:error] = errors
+      @users = User.all
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: remove

File: app/controllers/teams_controller.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  private
+
+  def team_params
+    params.require(:team).permit(:name, :description, :lead_id,
+                                 :sequence_num, :page)
```
Comment (bk-az) [inline app/controllers/teams_controller.rb]: sequence_num and page not required

File: app/controllers/time_logs_controller.rb [hunk 9]
```diff
@@ -20,6 +20,7 @@ def create
     @time_log.user_id = current_user.id
     if @time_log.save
       flash.now[:notice] = t 'time_log.created'
+      redirect_to @time_log
```
Comment (bk-az) [inline app/controllers/time_logs_controller.rb]: redirect_to @task

File: app/controllers/users_teams_controller.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    end
+  end
+
+  def destroy
+    if @users_team.destroy
+      flash[:notice] = t 'users_team.destroyed'
+    else
+      flash[:notice] = t 'users_team.not_destroyed'
+    end
+    redirect_to Team.find @users_team.team_id
```
Comment (bk-az) [inline app/controllers/users_teams_controller.rb]: redirect_to team_path(@users_team.team)

File: app/javascript/packs/team.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+require("datatables.net-bs4");
+document.currentScript.setAttribute("data-turbolinks-eval", "false");
```
Comment (bk-az) [inline app/javascript/packs/team.js]: remove

File: app/models/concerns/abilities/team_ability.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+module Abilities
+  # Abilities on team
+  module TeamAbility
+    def define_team_abilities_for_admin(user)
+      can :manage, Team, company_id: user.company_id
+    end
+
+    def define_team_abilities_for_employee(user)
+      can %i[read create], Team, company_id: user.company_id
+      can :update, Team, created_by_id: user.id, company_id: user.company_id
```
Comment (bk-az) [inline app/models/concerns/abilities/team_ability.rb]: remvove, also remove create

File: app/models/concerns/abilities/user_team_ability.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# Abilities
+module Abilities
+  # Abilities on team
+  module UserTeamAbility
+    def define_user_team_abilities_for_admin(user)
+      can :manage, UsersTeam, company_id: user.company_id
```
Comment (bk-az) [inline app/models/concerns/abilities/user_team_ability.rb:8]: restrict based on lead_id

File: app/models/concerns/abilities/user_team_ability.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    end
+
+    def define_user_team_abilities_for_employee(user)
+      can %i[read], UsersTeam, company_id: user.company_id
```
Comment (bk-az) [inline app/models/concerns/abilities/user_team_ability.rb:12]: can read if part of team

File: app/views/teams/_form.html.erb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+        <div class="form-group">
+          <%= form.label t(:team_name) %>
+          <%= form.text_field :name, class: 'form-control', required: true, minlength: 3 %>
+        </div>
+      </div>
+
+      <div class="col-lg-6">
+        <div class="form-group">
+          <%= form.label t(:team_lead_id) %>
+          <%= form.select :lead_id, options_for_select( @users.collect{ |u| [u.first_name,u.id] }), {}, {class: "form-control"} %>
```
Comment (bk-az) [inline app/views/teams/_form.html.erb]: token input

File: app/views/teams/_table.html.erb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    <th><%= t(:team_created_at) %></th>
+  </tr>
+  </thead>
+  <tbody>
+  
+  <% @teams.collect do |team| %>
+    <tr>
+      <td><%= link_to team.name, team_path(team) %> </td>
+      <td> <%= team.description %> </td>
+      <td> <%= link_to User.find(team.created_by_id).first_name + " " + User.find(team.created_by_id).last_name, user_path(User.find team.created_by_id) %> </td>
```
Comment (bk-az) [inline app/views/teams/_table.html.erb]: User.find(team.created_by_id).first_name + " " + User.find(team.created_by_id).last_name
team.created_by.full_name

Comment (bk-az) [inline app/views/teams/_table.html.erb]: user_path(User.find team.created_by_id)

user_path(User.find team.created_by)

File: app/views/teams/edit.html.erb [hunk 17]
```diff
@@ -0,0 +1,4 @@
+<div class="container-fluid">
+  <h1> <%= t(:team_update) %> </h1>
+</div>
+<%= render partial: "form",locals: {team: @team} %>
```
Comment (bk-az) [inline app/views/teams/edit.html.erb]: space

File: app/views/teams/new.js.erb [hunk 18]
```diff
@@ -0,0 +1,40 @@
+<div class="modal" id="users-modal">
+  <div class="modal-dialog">
+    <div class="modal-content">
+
```
Comment (bk-az) [inline app/views/teams/new.js.erb]: remove fle

File: app/views/teams/show.html.erb [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+        <th><%= t(:team_name) %></th>
+        <td><%= @team.name %></td>
+      </tr>
+      <tr>
+        <th><%= t(:team_description) %></th>
+        <td><%= @team.description %></td>
+      </tr>
+      <tr>
+        <th><%= t(:team_created_by) %></th>
+        <td><%= User.find(@team&.created_by_id).first_name + " " + User.find(@team&.created_by_id).last_name %></td>
```
Comment (bk-az) [inline app/views/teams/show.html.erb]: resolve this everywhare

File: app/models/concerns/abilities/user_team_ability.rb (hunk 13 above)
Comment (zhamza372) [reply app/models/concerns/abilities/user_team_ability.rb:8]: need to resolve this

File: app/controllers/tasks_controller.rb [hunk 20]
```diff
@@ -89,7 +89,7 @@ def destroy
   def task_params
     params.require(:task).permit(:title, :description, :start_date, :end_date,
                                  :due_date, :expected_start_date, :expected_end_date,
-                                 :priority, :status, :assignable_id,
-                                 :assignable_type, :progress, :sequence_num, :project_id, :page)
+                                 :priority, :status, :assignee_id,
+                                 :progress, :sequence_num, :project_id, :page)
```
Comment (bk-az) [inline app/controllers/tasks_controller.rb]: remove :sequence_num and :page

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -89,6 +89,15 @@ def destroy
     end
   end
 
+  def search
+    @users = User.where('first_name like :q or last_name like :q', q: "%#{params[:q]}%")
+    @users = @users.map { |u| { id: u.id, name: u.full_name } }
+    respond_to do |format|
+      format.html
```
Comment (bk-az) [inline app/controllers/users_controller.rb]: remove

File: app/views/tasks/_form.html.erb [hunk 22]
```diff
@@ -90,8 +90,8 @@
       
       <div class="col-lg-6">
         <div class="form-group">
-          <%= form.label t(:task_assignable_id) %>
-          <%= form.select :assignable_id, options_for_select( @users.collect{ |u| [u.first_name,u.id] }), {}, {class: "form-control"} %>
+          <%= form.label t(:task_assignee_id) %>
+          <%= form.select :assignee_id, options_for_select( @users.collect{ |u| [u.full_name,u.id] }), {}, {class: "form-control"} %>
```
Comment (bk-az) [inline app/views/tasks/_form.html.erb]: token input

**Pull request merged**